### PR TITLE
[OPIK-446] Add duration metric

### PR DESCRIPTION
**Pull request by thiagohora**
## Details
Add duration metric

Example payload:

```json
{
   "project_id":"0193bfa5-0387-7e9a-8720-abe3bb06cd2a",
   "metric_type":"DURATION",
   "interval":"HOURLY",
   "results":[
      {
         "name":"duration.p50",
         "data":[
            {
               "time":"2024-12-13T06:00:00Z",
               "value":null
            },
            {
               "time":"2024-12-13T07:00:00Z",
               "value":442.000000000
            },
            {
               "time":"2024-12-13T08:00:00Z",
               "value":null
            },
            {
               "time":"2024-12-13T09:00:00Z",
               "value":714.000000000
            },
            {
               "time":"2024-12-13T10:00:00Z",
               "value":303.000000000
            }
         ]
      },
      {
         "name":"duration.p99",
         "data":[
            {
               "time":"2024-12-13T06:00:00Z",
               "value":null
            },
            {
               "time":"2024-12-13T07:00:00Z",
               "value":934.999999999
            },
            {
               "time":"2024-12-13T08:00:00Z",
               "value":null
            },
            {
               "time":"2024-12-13T09:00:00Z",
               "value":805.720000000
            },
            {
               "time":"2024-12-13T10:00:00Z",
               "value":866.240000000
            }
         ]
      },
      {
         "name":"duration.p90",
         "data":[
            {
               "time":"2024-12-13T06:00:00Z",
               "value":null
            },
            {
               "time":"2024-12-13T07:00:00Z",
               "value":764.000000000
            },
            {
               "time":"2024-12-13T08:00:00Z",
               "value":null
            },
            {
               "time":"2024-12-13T09:00:00Z",
               "value":785.200000000
            },
            {
               "time":"2024-12-13T10:00:00Z",
               "value":814.400000000
            }
         ]
      }
   ]
}
```

## Issues
OPIK-446

### CODE REVIEW

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java [hunk 1]
```diff
@@ -3804,6 +3821,16 @@ private Span getAndAssert(Span expectedSpan, String apiKey, String workspaceName
             assertThat(actualSpan.lastUpdatedAt()).isAfter(expectedSpan.lastUpdatedAt());
             assertThat(actualSpan.createdBy()).isEqualTo(USER);
             assertThat(actualSpan.lastUpdatedBy()).isEqualTo(USER);
+            SoftAssertions.assertSoftly(softly -> {
```
Comment (idoberko2) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java]: There's only one assertion executed in each test run, I don't see why we need a soft assertion here

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java [hunk 2]
```diff
@@ -3384,6 +3390,17 @@ private void assertIgnoredFields(List<Span> actualSpans, List<Span> expectedSpan
                                     .build())
                     .isEqualTo(expectedFeedbackScores);
 
+            SoftAssertions.assertSoftly(softly -> {
```
Comment (idoberko2) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/SpansResourceTest.java]: There's only one assertion executed in each test run, I don't see why we need a soft assertion here

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/TracesResourceTest.java [hunk 3]
```diff
@@ -3353,6 +3358,18 @@ private static void assertIgnoredFields(Trace actualTrace, Trace expectedTrace)
         assertThat(actualTrace.lastUpdatedAt()).isAfter(expectedTrace.lastUpdatedAt());
         assertThat(actualTrace.createdBy()).isEqualTo(USER);
         assertThat(actualTrace.lastUpdatedBy()).isEqualTo(USER);
+
+        SoftAssertions.assertSoftly(softly -> {
```
Comment (idoberko2) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/TracesResourceTest.java]: Same comment about the soft assertion

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/TracesResourceTest.java [hunk 4]
```diff
@@ -3378,6 +3395,16 @@ private void assertIgnoredFieldsSpans(List<Span> actualSpans, List<Span> expecte
             assertThat(actualSpan.projectName()).isNull();
             assertThat(actualSpan.createdAt()).isAfter(expectedSpan.createdAt());
             assertThat(actualSpan.lastUpdatedAt()).isAfter(expectedSpan.lastUpdatedAt());
+            SoftAssertions.assertSoftly(softly -> {
```
Comment (idoberko2) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/TracesResourceTest.java]: Same comment about the soft assertion

**Review by idoberko2 — COMMENTED** [review]
LGTM. Left some minor comments about soft assertions. Feel free to merge and fix later

File: apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ProjectMetricsResourceTest.java [hunk 5]
```diff
@@ -362,11 +361,7 @@ public static Stream<Arguments> invalidParameters() {
                     arguments(named("start equal to end", validReq.toBuilder()
                             .intervalStart(now)
                             .intervalEnd(now)
-                            .build()), ProjectMetricsService.ERR_START_BEFORE_END),
-                    arguments(named("not supported metric", validReq.toBuilder()
-                            .metricType(MetricType.DURATION)
-                            .build()), ProjectMetricsService.ERR_PROJECT_METRIC_NOT_SUPPORTED.formatted(
-                                    MetricType.DURATION)));
```
Comment (andrescrz) [inline apps/opik-backend/src/test/java/com/comet/opik/api/resources/v1/priv/ProjectMetricsResourceTest.java:369]: What about this test case? The production code seems reachable.

**Review by andrescrz — APPROVED** [review]
Once tests pass, this LGTM.

**Review by andrescrz — APPROVED** [review]
LGTM, just the previous open question about that test.

**Comment by idoberko2** [comment]
@andrescrz, the removed test case is no longer relevant. It was covering the case that `MetricType` enum contains a value that doesn't have a corresponding metric implementation. Due to the changes in this PR, all enum values have corresponding implementations. I removed the dead code and wrote a proper test in: https://github.com/comet-ml/opik/commit/ca78abb3317683d941d5e6b1eec87cf20d29a1fa

**Pull request merged**